include/dlaf/matrix/internal/tile_future_manager.h

```diff
@@ -23,16 +23,21 @@ template <class ReturnTileType>
 pika::future<ReturnTileType> setPromiseTileFuture(
     pika::future<typename ReturnTileType::TileDataType> old_future,
     pika::lcos::local::promise<typename ReturnTileType::TileDataType> p) noexcept {
+  using pika::execution::experimental::keep_future;
+  using pika::execution::experimental::make_future;
+  using pika::execution::experimental::then;
+
   using TileDataType = typename ReturnTileType::TileDataType;
   using NonConstTileType = typename ReturnTileType::TileType;
 
   DLAF_ASSERT_HEAVY(old_future.valid(), "");
-  return old_future.then(pika::launch::sync, [p = std::move(p)](
-                                                 pika::future<TileDataType>&& fut) mutable {
+
+  // This uses keep_future because we want to handle exceptions in a special way
```

Comment:
Can you quickly explain this?

Reply:
This is partly to keep the current behaviour with futures and partly because I found it too awkward/impossible to express the same behaviour without `keep_future`.

If this didn't use `keep_future` exceptions would go through a separate channel and would not be handled by the lambda `set_promise`, and the promise would not get the exception set. Not necessarily wrong, but different, and I did not want to change the semantics at the moment.

A naive implementation could look a bit like:
```
return std::move(old_future) |
  let_error([&](std::exception_ptr ep) { p.set_exception(ep); }) |
  then([&](TileDataType&& tile) { return ReturnTileType(std::move(NonConstTileType(std::move(tile)).setPromise(std::move(p)))); });
```
but this does not handle the lifetime of the promise `p`. Seeing this again, it's possible this could be expressed with an additional `let_value` around this.

Alternatively, the semantics could also be changed, but I wouldn't want to do that here and without some additional discussion.